barretenberg/cpp/src/barretenberg/stdlib_circuit_builders/ultra_circuit_builder.hpp

```diff
@@ -452,10 +478,31 @@ class UltraCircuitBuilder_ : public CircuitBuilderBase<typename ExecutionTrace_:
     void create_unconstrained_gates(const std::vector<uint32_t>& variable_index);
 
     /**
-     * sort constraints for (batched) range checks.
+     * @brief Check for a sequence of variables that the neighborind differences are in {0, 1, 2, 3} via the delta_range
+     * block.
+     *
+     * @param variable_indices
+     */
+    void enforce_small_deltas(const std::vector<uint32_t>& variable_indices);
+    /**
+     * @brief Constrain consecutive variable differences to be in {0, 1, 2, 3}, _with_ boundary checks.
+     *
+     * @details Enforces that:
+     *   1. variable_indices[0] == start
+     *   2. variable_indices[i+1] - variable_indices[i] ∈ {0, 1, 2, 3} for all adjacent pairs
+     *   3. variable_indices[last] == end
+     *
+     * This is the core primitive for batched range checks: given a sorted list with bounded deltas
+     * starting at 0 and ending at N, all elements are proven to lie in [0, N].
+     *
+     * @param variable_indices The sequence of variable indices to constrain. Must have size > NUM_WIRES
+     *                         and divisible by NUM_WIRES (pad if necessary).
```

Comment:
Is there a good reason for this method requiring the input to be divisible by NUM_WIRES? Seems like a weird footgun to pass on to the user if the method could easily handle it internally as necessary.